Support for non-zero zero_points for weight and activation

dskhudia opened the pull request:
Summary: We can now have any valid zero points for weight and activation for conv2d kernel.

Differential Revision: D16556515

raghuramank100 commented on test/test_quantized.py:
```diff
@@ -467,46 +521,39 @@ def test_qconv(
 
-        # run on an input tensor
-        result_ref = conv_op(X_init.to(dtype=torch.float))
+        X_scale = 1.5
```
Can you add randomized testing for scale, zero-point values, so that we get more coverage?

dskhudia replied:
I am already getting time outs with the number of parameters I have.

dskhudia replied:
@raghuramank100: Current version of hypothesis (in internal code) doesn't allow me to increase deadline. In fact, I have to reduce the space as some of the tests are failing in OSS tests. Will update with reduced parameters.

zafartahirov replied:
Testing for scale and zero_points is important, IMHO. I honestly don't think that it's the hypothesis that is taking very long, but the test itself

raghuramank100 commented on test/test_quantized.py:
```diff
@@ -402,14 +456,14 @@ def test_qlinear_unpack(self, W):
 class TestQuantizedConv(unittest.TestCase):
     """Tests the correctness of quantized convolution op."""
-    @given(batch_size=st.integers(1, 3),
+    @given(batch_size=st.integers(1, 2),
```
What is the reason for reducing the amount of randomization? Is it the testing time?

dskhudia replied:
Yes. I was getting errors like the following:
> hypothesis.errors.Timeout: Ran out of time before finding a satisfying example for test_qconv. Only found 2 examples in 71.95s.

zafartahirov replied:
Your tests perform pretty heavy computation in python -- they will take a very long time.

dskhudia replied:
I agree the test itself is heavy.

raghuramank100 reviewed (changes requested):
Please address comments, thanks

zafartahirov commented on test/test_quantized.py:
```diff
@@ -66,6 +66,61 @@ def qlinear_ref(X_q, X_scale, X_zp, W_q, W_scale, W_zp, b_q, Y_scale, Y_zp):
+    # perform conv and calculate row offsets as well
+    result_ref = torch.zeros([N, OC, OH, OW], dtype=torch.float)
+    row_offsets_ref = torch.zeros([N, OC, OH, OW], dtype=torch.float)
+    for i in range(N):
```
The nested loops will make your test run long, and if you decide to use it in the hypothesis tests, it will almost certainly violate the deadline of 200ms. A better way is to use the `torch.nn.functional.conv2d` and do the quantization trick at the input/output

raghuramank100 replied:
Please use functional.conv2d, having zero-point map to float zero should work and give you close values, that you can compare with allclose.

dskhudia commented:
@raghuramank100 @zafartahirov : Please take a look at it now. I am using nn.Conv2d now and added random testing for scales/zero_points

dskhudia committed:
Support for non-zero zero_points for weight and activation (#23541)

Summary:
We can now have any valid zero points for weight and activation for conv2d kernel.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/23541

Test Plan:
buck test mode/dev caffe2/test:quantized -- 'test_qconv\ \(test_quantized.TestQuantizedConv\)'  --print-passing-details
```
Running 1 tests
Started new test run: https://our.intern.facebook.com/intern/testinfra/testrun/3377699723897843
      ✓ caffe2/test:quantized - test_qconv (test_quantized.TestQuantizedConv) 68.528 1/1 (passed)
Test output:
> test_qconv (test_quantized.TestQuantizedConv) ... ok
>
> ----------------------------------------------------------------------
> Ran 1 test in 68.529s
>
> OK
Finished test run: https://our.intern.facebook.com/intern/testinfra/testrun/3377699723897843
Summary (total time 74.97s):
  PASS: 1
  FAIL: 0
  SKIP: 0
  FATAL: 0
  TIMEOUT: 0
  OMIT: 0
```

Differential Revision: D16556515

Pulled By: dskhudia

fbshipit-source-id: 1d5ee9c4e925a7350003c423c9e5603f1abe03ea
